src/formatter.ts

```diff
@@ -691,14 +704,22 @@ export class Formatter {
       firstContext.getMetrics().totalLeftPx;
     let targetWidth = adjustedJustifyWidth;
     let actualWidth = shiftToIdealDistances(calculateIdealDistances(targetWidth));
-    const maxX = adjustedJustifyWidth + lastContext.getMetrics().notePx;
+    const musicFont = Flow.DEFAULT_FONT_STACK[0];
+    // The min padding we allow to the left of the right bar, before moving things left
+    const paddingMin = musicFont.lookupMetric('stave.endPaddingMin');
+    // the max padding we allow to the left of the right bar, before moving things back right.
+    const paddingMax = musicFont.lookupMetric('stave.endPaddingMax');
+    const maxX = adjustedJustifyWidth + lastContext.getMetrics().notePx - paddingMin;
 
     let iterations = this.formatterOptions.maxIterations;
-    while (actualWidth > maxX && iterations > 0) {
-      // If we couldn't fit all the notes into the jusification width, it's because the softmax-scaled
-      // widths between different durations differ across stave (e.g., 1 quarter note is not the same pixel-width
-      // as 4 16th-notes). Run another pass, now that we know how much to justify.
-      targetWidth -= actualWidth - maxX;
+    while ((actualWidth > maxX && iterations > 0) || (actualWidth + paddingMax < maxX && iterations > 1)) {
```

Comment:
`>1` because we always want to end with < actualWidth pixels, we don't want adjusting right as our last iteration.